### DISCUSSION TITLE
`mutate` not clearing errors

### DISCUSSION THREAD

**Discussion by samuliasmala** [vercel/swr#2491] [Ideas]
# Bug report

## Description / Observed Behavior

Calling `mutate` is not clearing errors when using `swr@2.0.0-beta7` or newer. Prior to that version (e.g., in `2.0.0-beta.6`) erros are cleared properly when calling `mutate`.

## Expected Behavior

I expect SWR to clear errors when calling mutate as it was doing previously. E.g., when clearing cache errors should be cleared with data:

```js
  const clearCache = () =>
    mutate(
      () => true, // Update all keys in cache
      undefined, // Update cache data to `undefined`
      { revalidate: false } // Do not revalidate
    );
```

## Repro Steps / Code Example

* [CodeSandbox](https://codesandbox.io/s/swr-mutate-clearing-errors-su2t80) where errors are cleared properly (`swr@2.0.0-beta6` )
* [CodeSandbox](https://codesandbox.io/s/swr-mutate-not-clearing-errors-lf33ke) where errors are not cleared (`swr@2.0.3` ).

`swr` version is the only difference between the two.

## Additional Context

Currently when using  `swr@2.0.3` all errors can be cleared with a following workaround:

```js
      const keys = cache.keys();
      if (keys) {
        let key = keys.next();
        while (key?.value) {
          cache.delete(key.value);
          key = keys.next();
        }
      }
```

**Comment by EdmundKorley** [comment]
Another workaround is to use the new [`useSWRMutation`] hook and call the [`reset`] method to reset the error state:

```jsx
import React from "react";
import useSWRMutation from "swr/mutation";

export default function ModalContent() {
  const { trigger, isMutating, error, reset } = useSWRMutation(
    "/api/foo",
    () => {
      throw new Error("Something went wrong");
    }
  );

  React.useEffect(() => {
    // This resets any error state from a previous useSWRMutation call
    reset();
  }, [reset]);

  return (
    <>
      <button disabled={isMutating} onClick={async () => await trigger()}>
        Create Something
      </button>
      {error && <span>{error.message}</span>}
    </>
  );
}

```

[`useSWRMutation`]: https://swr.vercel.app/docs/mutation#useswrmutation
[`reset`]: https://swr.vercel.app/docs/mutation#return-values-1

**Comment by samuliasmala** [comment]
I'm wondering why the issue was moved to discussions? I think undocumented change how errors are cleared at version `swr@2.0.0-beta7` should be either fixed or documented.

**Reply by koba04** [reply]
@samuliasmala Thank you for your feedback. Yeah, this is a breaking change, so we should have documented it in the v2 blog post as a breaking change even if we introduced it in the v2.0.0-rc.0 release.

https://github.com/vercel/swr/releases/tag/2.0.0-rc.0

The motivation is described in the PR.
https://github.com/vercel/swr/pull/2182

**Comment by pencil** [comment]
I ran into this bug as well and don't quite understand why this is expected behavior. In my case I was doing the following:

1. User opens `/loggedin` in their browser
1. Browser calls `useSWR<User>('/users/user/', authedFetcher(token))` to fetch the user data of the logged in user
2. Server responds with a 401 because the token is expired
3. User is sent to `/login` and prompted for login credentials
4. Browser sends credentials to server to authenticate
6. Server sends back `user` and `token`
7. Browser calls `mutate('/users/user', user)` to store newly received user data
8. User is sent to `/loggedin` again
9. Browser calls `useSWR<User>('/users/user/', authedFetcher(token))` again **but instead of just receiving the stored user data, the 401 error from the first call is also returned**

I fixed this by calling `cache.delete('/users/user/')` before calling `mutate('/users/user', user)` but I did not find this intuitive and I did not find this behavior documented anywhere.
